### DISCUSSION TITLE
`Mocked` type is missing in `@suites/unit`

### DISCUSSION THREAD

**Discussion by michaelcai** [automock/automock#473] [General]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

Description:
The @suites/unit package does not export the Mocked type as expected. When trying to import the Mocked type from @suites/unit using:
```typescript
import { TestBed, Mocked } from '@suites/unit';
```
I receive the following error message:
Module '"@suites/unit"' has no exported member named 'Mocked'.

### Minimum reproduction code

no

### Expected behavior

Please include the Mocked type in the package's exports, or provide documentation for an alternative way to mock objects if the Mocked type has been deprecated or replaced.

### Suites version

3.0.0

### Node.js version

20.16.0

### In which operating systems have you tested?

- [X] macOS
- [ ] Windows
- [ ] Linux

**Comment by omermorad** [comment]
Hi,

Thank you for submitting this issue. I'll take a look and update here soon!

**Comment by roniavi** [comment]
Plus one on this @omermorad 🙏

**Comment by qballer** [comment]
I will look into this issue in October and try and solve it.

**Comment by omermorad** [comment]
@roniavi @michaelcai - While @qballer works to resolve this problem, please try the following workaround:

Try importing `Mocked` from `@suites/doubles.x` instead of `@suites/unit`. (In this case, `@suites/doubles.jest`).

```typescript
import type { Mocked } from '@suites/doubles.jest';
```

This is a workaround; in order to maintain consistency and library agnostic behavior (mocking library), the `Mocked` type should essentially be imported from `@suites/unit`. You may easily modify it back to be imported from `@suites/unit` after the bug has been fixed.

Tell me whether it was successful please :)

**Comment by qballer** [comment]
https://github.com/qballer/suites_issue_426
I was able to reconstruct the bug in the above repo relatively easily. I will dig into why this is happening later.

**Comment by qballer** [comment]
I reconstructed the bug in the exact way described in the issue. 

https://github.com/qballer/suites_issue_426

It works out of the box as expected, but after installing an additional package (e.g `yarn add get-port`), for some reason, the following is missing:
in  
`/node_modules/@suites/unit/dist/cjs/index.d.ts`

Is missing:
```/// <reference types="@suites/doubles.jest/unit" />```

The same triple slash is missing in `node_modules/@suites/unit/dist/esm/index.d.ts`

Manually adding that to the d.ts file fixes the problem. 

I've also noticed that the following is missing from `node_modules/@suites/types.di/typings.d.ts`, which is 
 ```export type IdentifierMetadata = import('@suites/di.nestjs').IdentifierMetadata;```
 
 Digging through the code, these are all added via post-install scripts of `doubles.x` and `di.x`.  It looks like the fault is yarn for not running post-install script. this isn't happening in npm for example. 
 
Which yarn version is being used? 

https://github.com/yarnpkg/yarn/issues/5476

**Comment by omermorad** [comment]
@michaelcai @jeffrey-jian @roniavi (also related to #439 - @mbrunoviegas)

Due to a known issue with Yarn v1.x's `postinstall` script handling (https://github.com/yarnpkg/yarn/issues/5476), an extra step is required to set up reference types for your project. I've updated the docs, please follow: https://suites.dev/docs/overview/installation#for-yarn-v1x-users.

Please, let me know if it worked for you.

**Reply by bm77525-kr** [reply]
I'm using npm and `Mocked` is still not available for me from `@suites/unit`

**Reply by nhruch** [reply]
@omermorad This seems to also be an issue with PNPM. The provided workaround for yarn seems to have resolved the types correctly for me.

**Reply by jowusu837** [reply]
Worked for me, thanks!

**Reply by tomcastro** [reply]
Can confirm that the issue also happens in Yarn v4. Adding the global.d.ts file fixes another issue for me, that 

```ts
import { TestBed } from '@suites/unit';
import type { Mocked } from '@suites/doubles.vitest';

let messagesController: MessagesController;
let messagesService: Mocked<MessagesService>;

const { unit, unitRef } = await TestBed.solitary(MessagesController).compile();

messagesController = unit;
messagesService = unitRef.get(MessagesService); // ERROR HERE: Type 'StubbedInstance<MessagesService>' is not assignable to type 'Mocked<MessagesService>'.
```

I still have to import `Mocked` from `@suites/doubles.vitest` instead of `@suites/unit` though 😕

**Reply by omermorad** [reply]
> Can confirm that the issue also happens in Yarn v4. Adding the global.d.ts file fixes another issue for me, that
> 
> ```ts
> import { TestBed } from '@suites/unit';
> import type { Mocked } from '@suites/doubles.vitest';
> 
> let messagesController: MessagesController;
> let messagesService: Mocked<MessagesService>;
> 
> const { unit, unitRef } = await TestBed.solitary(MessagesController).compile();
> 
> messagesController = unit;
> messagesService = unitRef.get(MessagesService); // ERROR HERE: Type 'StubbedInstance<MessagesService>' is not assignable to type 'Mocked<MessagesService>'.
> ```
> 
> I still have to import `Mocked` from `@suites/doubles.vitest` instead of `@suites/unit` though 😕

@tomcastro Could you please let me know the location of the `global.d.ts` file? Could you please provide the folder structure or the repository for reproduction?

**Reply by tomcastro** [reply]
@omermorad It's a monorepo with yarn workspaces (and turborepo). Structure looks kinda like:

```
apps/
--app1/
--app2/
---- src/
------ mytest.spec.ts
---- global.d.ts
---- package.json (@suites/unit and @suites/doubles.vitest are installed here)
yarn.lock
package.json
```

**Reply by omermorad** [reply]
@tomcastro I see. It might be that the `global.d.ts` file needs to be placed in a different location within the monorepo. Could you please provide a minimum reproduction repository so I can play with it?

**Comment by omermorad** [comment]
Fixed in https://github.com/suites-dev/docs.suites.dev/commit/fde85a431392eef015bfd0affc4218f3ba01ab92, see https://suites.dev/docs/overview/installation#type-reference-configuration

@tomcastro - any updates?

**Comment by saurabh-git-dev** [comment]
Having similar issue with nx nestjs with pnpm project.
Same thing working when I use npm. Without any extra configuration.

Trying to add **global.d.ts** which also not working. all deps are installed. 
<img width="830" height="145" alt="image" src="https://github.com/user-attachments/assets/de1ec757-ef06-4f29-8a01-483daa2825b1" />
